nifi-nar-bundles/nifi-smb-bundle/nifi-smb-processors/src/main/java/org/apache/nifi/processors/smb/FetchSmb.java

```diff
@@ -0,0 +1,230 @@
+                    .build();
+
+    public static final Set<Relationship> relationships = Collections.unmodifiableSet(new HashSet<>(asList(
+            REL_SUCCESS,
+            REL_FAILURE,
+            REL_INPUT_FAILURE
+    )));
+
+    private static final List<PropertyDescriptor> PROPERTIES = asList(
+            REMOTE_FILE,
+            REMOTE_FILE_FIELD,
+            SMB_CLIENT_PROVIDER_SERVICE,
+            RECORD_READER
+    );
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        boolean shouldDeleteFlowFile = false;
+
+        final SmbClientProviderService clientProviderService =
+                context.getProperty(SMB_CLIENT_PROVIDER_SERVICE).asControllerService(SmbClientProviderService.class);
+
+        try (SmbClientService client = clientProviderService.getClient()) {
+
+            if (context.getProperty(RECORD_READER).isSet()) {
+                final RecordReaderFactory recordReaderFactory =
+                        context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+
+                try (InputStream inFlowFile = session.read(flowFile);
+                        RecordReader reader = recordReaderFactory.createRecordReader(flowFile, inFlowFile, getLogger())
+                ) {
+                    final String fieldName = context.getProperty(REMOTE_FILE_FIELD).getValue();
+                    Record record;
+                    while ((record = reader.nextRecord()) != null) {
+                        final String fileName = record.getAsString(fieldName);
+                        final FlowFile outFlowFile = session.create(flowFile);
+                        fetchAndTransfer(session, client, outFlowFile, fileName);
+                    }
+                    shouldDeleteFlowFile = true;
+                } catch (Exception e) {
+                    handleInputError(e, session, flowFile);
+                }
+            } else {
+                PropertyValue property = context.getProperty(REMOTE_FILE);
+                final String fileName = property.evaluateAttributeExpressions(flowFile).getValue();
+                fetchAndTransfer(session, client, flowFile, fileName);
+            }
+        } catch (Exception e) {
+            getLogger().error("Couldn't connect to smb due to "+ e.getMessage());
+            flowFile = session.putAttribute(flowFile, ERROR_CODE_ATTRIBUTE, getErrorCode(e));
+            flowFile = session.putAttribute(flowFile, ERROR_MESSAGE_ATTRIBUTE, e.getMessage());
+            session.transfer(flowFile, REL_FAILURE);
+        }
```

Comment:
In what circumstances can it happen? I think the REL_INPUT_FAILURE would be more adequate here too and rollback may also be needed.